libshaderc_spvc/src/spvc.cc

```diff
@@ -164,74 +174,370 @@ void shaderc_spvc_compiler_release(shaderc_spvc_compiler_t compiler) {
+    for (unsigned i = 0; i < 4; ++i) {
+      unsigned char c = (*p >> (i * 8)) & 0xff;
+      if (!c) return;
+      out.push_back(c);
     }
   }
+}
 
-  TRY_IF_EXCEPTIONS_ENABLED {
-    result->output = compiler->compile();
-    // An exception during compiling would crash (if exceptions off) or jump to
-    // the catch block (if exceptions on) so if we're here we know the compile
-    // worked.
-    result->status = shaderc_compilation_status_success;
-  }
-  CATCH_IF_EXCEPTIONS_ENABLED(...) {
-    result->status = shaderc_compilation_status_compilation_error;
-    result->messages = "Compilation failed.  Partial source:";
-    result->messages.append(compiler->get_partial_source());
+// Specialized for ID list.
+template <>
+void GetOp(const spv_parsed_instruction_t* inst, unsigned index,
+           std::vector<uint32_t>& out) {
+  if (index < inst->num_operands) {
+    const spv_parsed_operand_t& op = inst->operands[index];
+    assert(op.type == SPV_OPERAND_TYPE_ID);
+    const uint32_t* p = inst->words + op.offset;
+    std::copy(p, p + inst->num_operands - index, std::back_inserter(out));
   }
-  return result;
 }
+
+// Recursive template base case.
+void GetOpsRecur(const spv_parsed_instruction_t* inst, unsigned index) {}
+
+// Recursive template to do any number of GetOp().
+template <typename First, typename... Rest>
+void GetOpsRecur(const spv_parsed_instruction_t* inst, unsigned index,
+                 First& first, Rest&... rest) {
+  GetOp(inst, index, first);
+  GetOpsRecur(inst, index + 1, rest...);
+}
+
+// Start the recursive template.
+template <typename... Outs>
+void GetOps(const spv_parsed_instruction_t* inst, Outs&... outs) {
+  GetOpsRecur(inst, 0, outs...);
+}
+
+// This struct holds a SPIRV-Cross compiler object of type 'CompilerType,'
+// all the state needed while building the spirv_cross::ParsedIR object consumed
+// by the compiler, and a shaderc_spvc_compilation_result_t which indicates
+// the result of validation/parsing/compiling.
+template <class CompilerType>
+struct Compiler {
+  CompilerType* cross = nullptr;
+  spirv_cross::SPIRFunction* current_function = nullptr;
+  spirv_cross::SPIRBlock* current_block = nullptr;
+  spirv_cross::ParsedIR ir;
+  shaderc_spvc_compilation_result_t result =
+      nullptr;  // NOT freed by our destructor
+
+  void consume_validation_message(spv_message_level_t level, const char* src,
+                                  const spv_position_t& pos,
+                                  const char* message) {
+    assert(result);
+    result->messages.append(message);
+    result->messages.append("\n");
+  }
+
+  static spv_result_t parse_header(void* user_data, spv_endianness_t endian,
+                                   uint32_t magic, uint32_t version,
+                                   uint32_t generator, uint32_t id_bound,
+                                   uint32_t reserved) {
+    return ((Compiler*)user_data)
```

Comment:
static_cast<Compiler*>() instead of c-style cast?

Reply:
yep